src/slurm_plugin/clustermgtd.py

```diff
@@ -879,46 +906,227 @@ def _terminate_orphaned_instances(self, cluster_instances, ips_used_by_slurm):
+                args=(
+                    cluster_metrics,
+                    exceptions_list,
+                ),
             )
-        # Configure root logger
+        )
+
+    for subthread in subthreads:
+        subthread.start()
+
+
+def _monitor_and_restart_threads(subthreads, cluster_metrics, exceptions_list):
+    while True:
         try:
-            fileConfig(config.logging_config, disable_existing_loggers=False)
-        except Exception as e:
-            log.warning(
-                "Unable to configure logging from %s, using default logging settings.\nException: %s",
-                config.logging_config,
-                e,
-            )
-        # Manage cluster
-        cluster_manager.manage_cluster()
-        sleep_remaining_loop_time(config.loop_time, start_time)
+            start_time = datetime.now(tz=timezone.utc)
+            if exceptions_list:
+                exception = exceptions_list.pop()
+                log.error("An unexpected error occurred: %s", exception)
+                raise exception
+            sleep_remaining_loop_time(cluster_metrics.config.loop_time // 2, start_time)
+        except ClusterMgtdException:
+            # restart _run_clustermgtd thread
+            log.info("Restarting _run_clustermgtd thread")
+            subthreads[0] = subthreads[0].clone()
+            subthreads[0].start()
+        except ClusterMetricsException:
+            # restart _push_metrics_to_cloudwatch thread
+            log.info("Restarting _push_metrics_to_cloudwatch thread")
+            subthreads[1] = subthreads[1].clone()
+            subthreads[1].start()
+
+
+def _create_and_monitor_threads(subthreads, clustermgtd_config_file, cluster_metrics, exceptions_list):
+
+    try:
+        _create_and_start_threads(subthreads, clustermgtd_config_file, cluster_metrics, exceptions_list)
+
+        _monitor_and_restart_threads(subthreads, cluster_metrics, exceptions_list)
+
+    except Exception as e:
+        log.error("An unexpected error occurred: %s", e)
+        raise CriticalThreadException()
 
 
 @retry(wait_fixed=seconds(LOOP_TIME))
 def main():
     logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s [%(module)s:%(funcName)s] %(message)s")
     log.info("ClusterManager Startup")
     try:
-        _run_clustermgtd()
+        cluster_metrics = ClusterMetrics()
+        clustermgtd_config_file = os.path.join(CONFIG_FILE_DIR, "parallelcluster_clustermgtd.conf")
+        config = ClustermgtdConfig(clustermgtd_config_file)
+        cluster_metrics.config = config
+        exceptions_list = []
+
+        subthreads = []
+
+        try:
+            _create_and_monitor_threads(subthreads, clustermgtd_config_file, cluster_metrics, exceptions_list)
+        except CriticalThreadException:
+            _terminate_threads(subthreads)
+
     except Exception as e:
-        log.exception("An unexpected error occurred: %s", e)
+        log.error("An unexpected error occurred: %s", e)
```

Comment:
in this case we want to log it as an exception since it's something unexpected and the stack trace is useful in this situation

Reply:
Fixed in the last force push, should work now even for children threads (changed the approach to log those with the trace)